### ISSUE TITLE
GPU adaptation problem

### ISSUE THREAD

**Issue by flamingoXu** [QuantumLab-ZY/HamGNN#14]
When I tried to run HamGNN in the GPU platform, I encountered some problem due to the version of the python library. However, the library that I have installed is completely in accordance with the readme requirements.
my GPU tpye is V100，pytorch（cuda11.3）
![691f2166ce538dcf691b793f9d900a7](https://github.com/QuantumLab-ZY/HamGNN/assets/152261627/e0494ff8-2e6f-4937-998b-2bf575275cb7)

![367492cb3d3055f9385649d6f24ff80](https://github.com/QuantumLab-ZY/HamGNN/assets/152261627/a9dd5851-e739-462e-a717-4ad70f958e1b)

**Comment by QuantumLab-ZY** [comment]
I want to check your config.yaml file.

**Comment by flamingoXu** [comment]
here is my config.yaml,and i also want to know whether this program can predict wavefunction? If could ,how to output Hamiltonian and Wavefunction together
dataset_params:
  batch_size: 1
  split_file: null
  test_ratio: 0.1
  train_ratio: 0.8
  val_ratio: 0.1
  graph_data_path: ./Examples/Moire_twisted_bilayer_MoS2/ # Directory where graph_data.npz is located

losses_metrics:
  losses:
  - loss_weight: 1.0
    metric: mae
    prediction: hamiltonian
    target: hamiltonian
  #- loss_weight: 1.0
  #  metric: mae
  #  prediction: band_gap
  #  target: band_gap
  #- loss_weight: 0.001
  #  metric: mae
  #  prediction: band_energy
  #  target: band_energy
  #- loss_weight: 1.0
  #  metric: mae
  #  prediction: overlap
  #  target: overlap
  #- loss_weight: 1.0
  #  metric: mae
  #  prediction: peak
  #  target: peak
  #- loss_weight: 0.0
  #  metric: mae
  #  prediction: hamiltonian_imag
  #  target: hamiltonian_imag
  #- loss_weight: 0.0001
  #  metric: abs_mae
  #  prediction: wavefunction
  #  target: wavefunction
  metrics:
  - metric: mae
    prediction: hamiltonian
    target: hamiltonian
  #- metric: mae
  #  prediction: band_gap
  #  target: band_gap
  #- metric: mae
  #  prediction: peak
  #  target: peak
  #- metric: mae
  #  prediction: overlap
  #  target: overlap
  #- metric: mae
  #  prediction: hamiltonian_imag
  #  target: hamiltonian_imag
  #- metric: mae
  #  prediction: hamiltonian_imag
  #  target: hamiltonian_imag
  #- metric: mae
  #  prediction: band_energy
  #  target: band_energy
  #- metric: abs_mae
  #  prediction: wavefunction
  #  target: wavefunction

# Generally, the optim_params module only needs to set the initial learning rate (lr)
optim_params:
  lr: 0.001
  lr_decay: 0.5
  lr_patience: 5
  gradient_clip_val: 0.0
  max_epochs: 3000
  min_epochs: 100
  stop_patience: 30

output_nets:
  output_module: HamGNN_out
  HamGNN_out:
    ham_only: true # true: Only the Hamiltonian H is computed; 'false': Fit both H and S
    ham_type: openmx # openmx: fit openmx Hamiltonian; abacus: fit abacus Hamiltonian
    nao_max: 19 # The maximum number of atomic orbitals in the data set, which can be 14, 19 or 27
    add_H0: true # Generally true, the complete Hamiltonian is predicted as the sum of H_scf plus H_nonscf (H0)
    symmetrize: true # if set to true, the Hermitian symmetry constraint is imposed on the Hamiltonian
    calculate_band_energy: false # Whether to calculate the energy bands to train the model
    num_k: 5 # When calculating the energy bands, the number of K points to use
    band_num_control: null # `dict`: controls how many orbitals are considered for each atom in energy bands; `int`: [vbm-num, vbm+num]; `null`: all bands
    k_path: null # `auto`: Automatically determine the k-point path; `null`: random k-point path; `list`: list of k-point paths provided by the user
    soc_switch: false # if true, fit the SOC Hamiltonian
    nonlinearity_type: norm # norm or gate

profiler_params:
  progress_bar_refresh_rat: 1
  train_dir: ./Examples/MoS2_tw_5.09 #The folder for saving training information and prediction results. This directory can be read by tensorboard to monitor the training process.

representation_nets:
  # Network parameters usually do not need to be changed.
  HamGNN_pre:
    cutoff: 20.0
    resnet: True
    cutoff_func: cos
    edge_sh_normalization: component
    edge_sh_normalize: true
    ######## Irreps set 1 (crystal): ################
    feature_irreps_hidden: 32x0o+32x0e+32x1o+32x1e+32x2e+32x2o+32x3o+32x3e+32x4o+32x4e
    irreps_edge_output: 32x0o+32x0e+32x1o+32x1e+32x2e+32x2o+32x3o+32x3e+32x4o+32x4e
    irreps_edge_sh: 0e + 1o + 2e + 3o + 4e
    irreps_node_features: 32x0o+32x0e+32x1o+32x1e+32x2e+32x2o+32x3o+32x3e+32x4o+32x4e
    irreps_node_output: 32x0o+32x0e+32x1o+32x1e+32x2e+32x2o+32x3o+32x3e+32x4o+32x4e
    irreps_triplet_output: 32x0o+32x0e+32x1o+32x1e+32x2e+32x2o+32x3o+32x3e+32x4o+32x4e
    invariant_layers: 2
    invariant_neurons: 64
    num_interaction_layers: 5
    num_radial: 8
    num_spherical: 8
    num_types: 60
    export_triplet: false
    rbf_func: bessel
    set_features: true
    add_edge_tp: false
    irreps_node_prev: 16x0o+16x0e+8x1o+8x1e+8x2e+8x2o+8x3o+8x3e+8x4o+8x4e
    num_node_attr_feas: 64

setup:
  GNN_Net: HamGNN_pre
  accelerator: null
  ignore_warnings: true
  checkpoint_path: ./network_weights_bilayer_MoS2.ckpt # Path to the model weights file
  load_from_checkpoint: false
  resume: false
  num_gpus: [1] # null: use cpu; [i]: use the ith GPU device
  precision: 32
  property: hamiltonian
  stage: test # fit: training; test: inference

**Comment by QuantumLab-ZY** [comment]
The problem seems to be caused by the pymatgen library, which throws this exception when executing 'from Pymatgen.core. structure import Structure'. The version of pymatgen I used is Pymatgen-2022.3.7, yours seems to be the latest version of 2024, you can try my version of pymatgen to see if the bug still exists.

> Dear Zhong Yang When I tried to run HamGNN in the GPU platform, I encountered some problem due to the version of the python library. However, the library that I have installed is completely in accordance with the readme requirements. my GPU tpye is V100，pytorch（cuda11.3） ![691f2166ce538dcf691b793f9d900a7](https://private-user-images.githubusercontent.com/152261627/325455879-e0494ff8-2e6f-4937-998b-2bf575275cb7.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwMjgwOTgsIm5iZiI6MTcxNDAyNzc5OCwicGF0aCI6Ii8xNTIyNjE2MjcvMzI1NDU1ODc5LWUwNDk0ZmY4LTJlNmYtNDkzNy05OThiLTJiZjU3NTI3NWNiNy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNVQwNjQ5NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00OWQxN2Q4MDIzYTBiYTc5ZDgyMjRmOTA4MzFhMjI5OTY3NzA5MjdjNzlhNmViYjdlN2ZhODM1OTZiMjAxYWVlJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.758wGeP_Ks8MdUvXL7pV6-4QVBBfNA1nJkz9DVhp2BQ)
> 
> ![367492cb3d3055f9385649d6f24ff80](https://private-user-images.githubusercontent.com/152261627/325455908-a9dd5851-e739-462e-a717-4ad70f958e1b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTQwMjgwOTgsIm5iZiI6MTcxNDAyNzc5OCwicGF0aCI6Ii8xNTIyNjE2MjcvMzI1NDU1OTA4LWE5ZGQ1ODUxLWU3MzktNDYyZS1hNzE3LTRhZDcwZjk1OGUxYi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNDI1JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDQyNVQwNjQ5NThaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1kZGYyOTljYzUzNGY5NDJhMGJkNWI3OWZiYjJhOGFiYmJhZWFkNDNlYzkyZTQxODhjYTI0OWZjMTZjM2I4NTFjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.nltnyXxbUZ9EOkLYc-G1QuBaSsXLG54YWAjY_-0nTOs)

**Comment by QuantumLab-ZY** [comment]
I have uploaded my HamGNN conda environment on this [website](https://zenodo.org/records/11064223), and users can directly extract this conda environment to their own `conda/envs` path.

**Comment by flamingoXu** [comment]
Dear Zhong Yang：
I had tried to change my pymatgen version into 22.3.7, but it doesn't work.  Then, I changed my networkx with an old version 2.8.8, which finally fixed the bug. 
It will be nice if you can label your python libs version in the readme file. which will help the user a lot.
Best wishes.

**Comment by QuantumLab-ZY** [comment]
This is a good suggestion, I will provide more detailed information about the Python environment when I have convenient time.

> Dear Zhong Yang： I had tried to change my pymatgen version into 22.3.7, but it doesn't work. Then, I changed my networkx with an old version 2.8.8, which finally fixed the bug. It will be nice if you can label your python libs version in the readme file. which will help the user a lot. Best wishes.

**Comment by JaGeo** [comment]
> This is a good suggestion, I will provide more detailed information about the Python environment when I have convenient time.
> 
> > Dear Zhong Yang： I had tried to change my pymatgen version into 22.3.7, but it doesn't work. Then, I changed my networkx with an old version 2.8.8, which finally fixed the bug. It will be nice if you can label your python libs version in the readme file. which will help the user a lot. Best wishes.

This would be really great for reuse! If you are interested in reuse, maybe an easier installation process via pip install and pypi or conda would also help! Thanks in advance!

**Comment by QuantumLab-ZY** [comment]
Dear JaGeo,

I appreciate your suggestion regarding an easier installation process. Unfortunately, this code cannot be installed simply via pip install due to its dependency on numerous external third-party libraries, which in turn rely on some additional Python packages with which I am not familiar. Resolving all these dependencies is quite challenging.

As a temporary solution, I have uploaded the Conda environment I am using to this [repository](https://zenodo.org/records/11064223). You can extract this Conda environment to your own `conda/envs` directory and then use the command `pip list --format=freeze > requirements.txt` within this Conda environment to generate a list of all the third-party Python dependency versions. This should help you get started more easily. If you have any suggestions for a better and simpler way to handle the environment dependencies for HamGNN, I would greatly appreciate it.

Thank you for your understanding and support!

Best regards,
Yang Zhong

> > This is a good suggestion, I will provide more detailed information about the Python environment when I have convenient time.
> > > Dear Zhong Yang： I had tried to change my pymatgen version into 22.3.7, but it doesn't work. Then, I changed my networkx with an old version 2.8.8, which finally fixed the bug. It will be nice if you can label your python libs version in the readme file. which will help the user a lot. Best wishes.
> 
> This would be really great for reuse! If you are interested in reuse, maybe an easier installation process via pip install and pypi or conda would also help! Thanks in advance!

**Comment by JaGeo** [comment]
@QuantumLab-ZY Thank you very much! This should already help quite a lot!
